notebooks/query-language-object-based-dialect/01.01.quantifier_capabilities.ipynb

```diff
@@ -0,0 +1,842 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "373ad780",
+   "metadata": {},
+   "source": [
+    "The following query matches all single nodes where the predicate that the `name` metric `starts with \"MPI_Barrier\"` is satisfied:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "455adb46",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# single node with names starting with MPI_Barrier\n",
+    "query_1 = [\n",
+    "    (1, {\"name\": \"MPI_Barrier.*\"}),\n",
+    "]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "82a5a9c6",
+   "metadata": {},
+   "source": [
+    "The above query, when used with an accepted value for the quantifier, \".\", to match all single nodes with a certain predicate, is equivalent to the follwing query:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a1adf78c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# single node with names starting with MPI_Barrier\n",
+    "query_1 = [\n",
+    "    (\".\", {\"name\": \"MPI_Barrier.*\"}),\n",
+    "]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e409873f-7736-4d08-8516-84c839a2229d",
+   "metadata": {},
+   "source": [
+    "The above query is passed to Hatchet’s `filter()` function to filter the input GraphFrame. The `filter()` function takes a user-supplied function or query object and applies that to all rows in the DataFrame. The resulting Series or DataFrame is used to filter the DataFrame to only return rows that are true."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "878ec2a0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "gf_filt = gf.filter(query_1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2040591a-06f6-462f-8528-bcfd0fedd22b",
+   "metadata": {},
+   "source": [
+    "The resulting GraphFrame now only lists the  nodes that matched the query:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7ef7369a-1968-43a6-b53a-a01b2a4785fe",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(gf_filt.tree())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "904d2a82-0f48-4071-b015-8f9bd1fc8d89",
+   "metadata": {},
+   "source": [
+    "The query match is also reflected in the DataFrame:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "082afe16",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "gf_filt.dataframe"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "579d1c97",
+   "metadata": {},
+   "source": [
+    "Then, we use  `drop_index_levels()`  to get rid of all index levels besides the node column in preparation of defining more queries below."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a36d95f9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "gf.drop_index_levels()"
```

Comment:
Is this not redundant to [line 167](https://github.com/LLNL/hatchet-tutorial/pull/4/files#diff-501eaa9d95f9c174c99c81d3917b4b7e8b859d2a3b1dc5dfe4866476490c8b16R167)? I'm confused why it is necessary to drop the index level again (as well as the subsequent 4 `gf.drop_index_levels()` calls).

Reply:
It really isn't necessary. Let me remove those redundancies.